helix-view/src/clipboard/config.rs

```diff
@@ -0,0 +1,248 @@
+                primary_paste: "wl-paste", "-p", "--no-newline";
+                env: "WAYLAND_DISPLAY";
+            }
+
+            #[cfg(unix)]
+            "xclip" {
+                copy: "xclip", "-i", "-selection", "clipboard";
+                paste: "xclip", "-o", "-selection", "clipboard";
+                primary_copy: "xclip", "-i";
+                primary_paste: "xclip", "-o";
+                env: "DISPLAY";
+            }
+
+            #[cfg(unix)]
+            "xsel" {
+                copy: "xsel", "-o", "-b";
+                paste: "xsel", "-i", "-b";
+                primary_copy: "xsel", "-o";
+                primary_paste: "xsel", "-i";
+                env: "DISPLAY";
+                test: "xsel", "-o", "-b";
+            }
+
+            "lemonade" {
+                copy: "lemonade", "copy";
+                paste: "lemonade", "paste";
+            }
+
+            #[cfg(unix)]
+            "doit" {
+                copy: "doitclient", "wclip";
+                paste: "doitclient", "wclip", "-r";
+            }
+
+            #[cfg(any(windows, target_os = "linux"))] // this is a godsend on WSL
+            "win32yank" {
+                copy: "win32yank.exe", "-i";
+                paste: "win32yank.exe", "-o";
+            }
+
+            #[cfg(unix)]
+            "tmux" {
+                copy: "tmux", "load-buffer", "-";
+                paste: "tmux", "save-buffer", "-";
+                env: "TMUX";
+            }
+
+            #[cfg(target_os = "linux")]
+            "termux" {
+                copy: "termux-clipboard-set";
+                paste: "termux-clipboard-get";
+            }
+
+            #[cfg(target_os = "macos")]
+            "pbcopy+pbpaste" {
+                copy: "pbcopy";
+                paste: "pbpaste";
+            }
+        };
+        if cfg!(windows) {
+            cfg.order.push("clipboard-win".to_owned());
+        }
+        cfg
+    }
+}
+
+impl ClipboardConfig {
+    pub fn get_provider(&self) -> Box<dyn ClipboardProvider> {
+        for name in &self.order {
+            #[cfg(windows)]
+            if name == "clipboard-win" {
+                return Box::new(provider::WindowsProvider);
+            }
+            if let Some(spec) = self.providers.get(name) {
+                if spec.is_available() {
+                    return Box::new(provider::CommandProvider::from(spec));
+                }
+            } else {
+                warn!("No clipboard provider named {name}");
+                continue;
+            }
+        }
+
+        Box::new(provider::NopProvider::new())
+    }
+}
+
+impl ClipboardProviderSpec {
+    pub fn is_available(&self) -> bool {
+        macro_rules! check {
+            ($cond:expr) => {
+                if !$cond {
+                    return false;
+                }
+            };
+        }
+
+        check!(!self.copy.is_empty() && exists(&self.copy[0]));
+        check!(!self.paste.is_empty() && exists(&self.paste[0]));
```

Comment:
I think this macro is unnecessary, it's just a couple of lines.

Reply:
I generally follow the rule of three for avoiding repetitive code. This is five (or four, without `test_commands`), and I'm not a fan of how it looks without the macro:
```rs
pub fn is_available(&self) -> bool {
    if !self.copy.is_empty() && !exists(&self.copy[0]) {
        return false;
    }
    if !self.paste.is_empty() && !exists(&self.paste[0]) {
        return false;
    }

    if let Some(copy) = &self.primary_copy {
        if !copy.is_empty() && !exists(&copy[0]) {
            return false;
        }
    }
    if let Some(paste) = &self.primary_paste {
        if !paste.is_empty() && !exists(&paste[0]) {
            return false;
        }
    }
    for var in &self.env_vars {
        if !env_var_is_set(var) {
            return false;
        }
    }
    for cmd in &self.test_commands {
        // FIXME: check performance of is_exit_success
        if !cmd.is_empty() && !is_exit_success(&cmd[0], &cmd[1..]) {
            return false;
        }
    }

    true
}
```